File: app/src/main/java/com/example/rickmortymvvm/list/viewmodel/ViewModelImpl.java
package com.example.rickmortymvvm.list.viewmodel;

import android.util.Log;

import androidx.annotation.NonNull;

import com.example.rickmortymvvm.list.services.CharacterService;
import com.example.rickmortymvvm.util.observer.MutableObservable;
import com.example.rickmortymvvm.util.observer.Observable;

import java.util.ArrayList;
import java.util.List;

import com.example.rickmortymvvm.models.Character;
import com.example.rickmortymvvm.list.services.models.CharacterResponseVO;
import com.example.rickmortymvvm.util.observer.MutableObservableImpl;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import static com.example.rickmortymvvm.list.viewmodel.PresentationCharacterListAction.*;

public class ViewModelImpl implements ViewModel {

    private int page = 1;
    private int pages = 1;
    private boolean requestAvailable = true;

    private Retrofit rf = new Retrofit
            .Builder()
            .baseUrl(CharacterService.BASE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build();
    private CharacterService service = rf.create(CharacterService.class);
    private MutableObservable mutableObservable = new MutableObservableImpl();
    private List<Character> list = new ArrayList<>();

    @Override
    public void onCreate() {
        requestCharacterList();
    }

    @Override
    public void onClickCharacter(Character character) {
        mutableObservable.update(new GoToInfo(character));
    }

    @Override
    public void onClickTryAgain() {
        requestCharacterList();
    }

    @Override
    public void onClickQuit() {
        mutableObservable.update(new Finish());
    }



    public Observable getObservable() {
        return mutableObservable;
    }

    @Override
    public void onScrollFinal() {
        requestCharacterList();
    }


    private void requestCharacterList() {
        final String TAG = "service";
        if (page <= pages && requestAvailable) {
            requestAvailable = false;
            mutableObservable.update(new PresentationCharacterListState(true, list, true, false));
            service.listCharacter(page)
                    .enqueue(
                            new Callback<CharacterResponseVO>() {
                                @Override
                                public void onResponse(@NonNull Call<CharacterResponseVO> call, @NonNull Response<CharacterResponseVO> response) {
                                    CharacterResponseVO characterResponse;

                                    if (response.isSuccessful()
                                            && (characterResponse = response.body()) != null
                                    ) {
                                        list.addAll(characterResponse.getResults());
                                        mutableObservable.update(new PresentationCharacterListState(false, list, true, false));
                                        pages = characterResponse.getInfo().getPages();
                                        page++;
                                    } else {
                                        Log.i(TAG, "Error: " + response.code());

                                        mutableObservable.update(new PresentationCharacterListState(false, list, false, true));
                                    }
                                    requestAvailable = true;
                                }
                                @Override
                                public void onFailure(@NonNull Call<CharacterResponseVO> call, @NonNull Throwable t) {
                                    Log.e(TAG, "Error : " + t.getMessage());
                                    mutableObservable.update(new PresentationCharacterListState(false, list, false, true));
                                    requestAvailable = true;
                                }
                            });
        }
    }
}

File: app/src/main/java/com/example/rickmortymvvm/list/viewmodel/ViewModelImpl.kt
package com.example.rickmortymvvm.list.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.rickmortymvvm.list.services.CharacterService
import com.example.rickmortymvvm.list.viewmodel.PresentationCharacterListAction.Finish
import com.example.rickmortymvvm.list.viewmodel.PresentationCharacterListAction.GoToInfo
import com.example.rickmortymvvm.models.Character
import com.example.rickmortymvvm.util.observer.MutableAction
import com.example.rickmortymvvm.util.observer.MutableMyObservable
import com.example.rickmortymvvm.util.observer.MutableMyObservableImpl
import com.example.rickmortymvvm.util.observer.MyObservable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

class ViewModelImpl : MyViewModel, ViewModel() {

    private var page = 1
    private var pages = 1
    private var requestAvailable = true
    private val rf = Retrofit.Builder()
        .baseUrl(CharacterService.BASE_URL)
        .addConverterFactory(GsonConverterFactory.create())
        .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
        .build()
    private val service = rf.create(CharacterService::class.java)
    private val mutableObservable: MutableMyObservable = MutableMyObservableImpl()
    private var list: MutableList<Character> = ArrayList()

    private val _action = MutableAction<PresentationCharacterListAction>()
    val action :LiveData<PresentationCharacterListAction>
        get() = _action

    private val _status = MutableLiveData(
        PresentationCharacterListState(true, list, false, false)
    )
    val state: LiveData<PresentationCharacterListState>
        get() = _status

    override fun onCreate() {
        if (list.size == 0) {
            requestCharacterList()
        }
    }

    override fun onClickCharacter(character: Character?) {
        _action.sendAction( GoToInfo(character!!))
    }

    override fun onClickTryAgain() {
        requestCharacterList()
    }

    override fun onClickQuit() {
        _action.value = Finish
    }

    override val myObservable: MyObservable
        get() = mutableObservable

    override fun onScrollFinal() {
        requestCharacterList()
    }

    private fun requestCharacterList() {
        val TAG = "service"
        if (page <= pages && requestAvailable) {
            requestAvailable = false
            _status.value = PresentationCharacterListState(true, list, true, false)
            val disposable = service.listCharacter(page)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response ->
                    val results = response?.results
                    if (results != null) {
                        list.addAll(response.results)
                        _status.value = PresentationCharacterListState(
                            false,
                            list,
                            true,
                            false
                        )
                        pages = response.info!!.pages
                        page++
                        requestAvailable = true
                    }
                }) {
                    _status.value = PresentationCharacterListState(
                        false,
                        list,
                        false,
                        true
                    )
                    requestAvailable = true
                }

        }
    }
}






